isofit/core/isofit.py

```diff
@@ -111,77 +119,130 @@ def run(self, row_column=None):
             ranges = row_column.split(",")
             if len(ranges) == 1:
                 self.rows, self.cols = [int(ranges[0])], None
             if len(ranges) == 2:
                 row_start, row_end = ranges
-                self.rows, self.cols = range(int(row_start), int(row_end)), None
+                self.rows = range(int(row_start), int(row_end))
+                self.cols = None
             elif len(ranges) == 4:
                 row_start, row_end, col_start, col_end = ranges
                 self.rows = range(int(row_start), int(row_end) + 1)
                 self.cols = range(int(col_start), int(col_end) + 1)
+
+        # Else running all of the file
         else:
-            io = IO(self.config, fm)
+            io = IO(self.config, self.full_statevector)
             self.rows = range(io.n_rows)
             self.cols = range(io.n_cols)
             del io
 
+        # Form the row-column pairs (pixels to run)
         index_pairs = np.vstack(
             [x.flatten(order="f") for x in np.meshgrid(self.rows, self.cols)]
         ).T
 
-        n_iter = index_pairs.shape[0]
-
-        if self.config.implementation.n_cores is None:
-            n_workers = multiprocessing.cpu_count()
-        else:
-            n_workers = self.config.implementation.n_cores
-
-        # Max out the number of workers based on the number of tasks
-        n_workers = min(n_workers, n_iter)
-
-        params = [
-            ray.put(obj)
-            for obj in [self.config, fm, self.loglevel, self.logfile, n_workers]
-        ]
-        self.workers = ray.util.ActorPool(
-            [Worker.remote(*params, n) for n in range(n_workers)]
-        )
-
-        start_time = time.time()
-        n_tasks = min(
-            n_workers * self.config.implementation.task_inflation_factor, n_iter
-        )
+        # Save this for logging
+        total_samples = index_pairs.shape[0]
 
-        logging.info(
-            f"Beginning {n_iter} inversions in {n_tasks} chunks using {n_workers} cores"
+        # If multistate, split into class. Flag is messy because of examples
+        index_pairs = index_spectra_by_surface(
+            self.config.forward_model.surface, index_pairs
         )
 
-        # Divide up spectra to run into chunks
-        index_sets = np.linspace(0, n_iter, num=n_tasks, dtype=int)
-        if len(index_sets) == 1:
-            indices_to_run = [index_pairs[0:1, :]]
+        # Some logging that might be nice
+        if len(index_pairs.keys()) > 1:
+            logging.info("Multi-state inversion started.")
         else:
-            indices_to_run = [
-                index_pairs[index_sets[l] : index_sets[l + 1], :]
-                for l in range(len(index_sets) - 1)
+            logging.info("Single-state inversion started.")
+
+        # Loop through index pairs and run workers
+        class_loop_start_time = time.time()
+        for surface_class_str, index_pair in index_pairs.items():
```

Comment:
Iterates through key, values of surface state and pixel row-col list. Updates the config to reflect the specific surface state, initializes the forward model, and parallel runs per pixel.